paper.tex

```diff
@@ -329,6 +306,114 @@ \subsection{Unrolled recursion}
+
+\subsection{Path encoding}
+
+The tag byte is a number of bit fields, packed into a byte.
+
+TODO: make this into a diagram.
+
+\begin{itemize}
+    \item 0-1: coordinate count
+    \item 2: subpath end bit
+    \item 3: 16/32 bit coordinates
+    \item 4: path end bit
+    \item 5: transform bit
+    \item 6: line style bit
+\end{itemize}
+
+For most path segments, the coordinate count indicates the type of path segment: 1 for lines, 2 for quadratic Beziers, and 3 for cubic Beziers. The subpath bit is set on the last segment of a subpath, and the path bit is additionally set on the last bit of a path. The latter is not directly interpreted by the path flattening code, but is forwarded to later rendering stages, for example, the path id can be used to select a color or other paint style.
+
+The size is readily inferred from the tag byte: the number of coordinate pairs is the coordinate count plus one if the subpath end bit is set. The nuber of bytes per coordinate pair is 4 for 16 bit coordinates or 8 for 32 bit coordinates.
+
+It is relatively straightforward to convert the standard moveto/lineto/curveto path representation into this encoding. Subpath boundaries are moved from the beginning of a subpath (moveto) to the end. A separate segment must be generated for closepath if the ending point does not coincide with the start point.
+
+The handling of the subpath bit allows for overlap in coordinates, so that, except at subpath boundaries, the first coordinate pair of each path segment overlaps with the last coordinate pair of the previous one. Subpath handling is all computed in the prefix sum; no special handling is needed on the GPU side.
+
+\subsection{Algorithm}
+
+% TODO: explain flatten_euler
+
+\subsection{Limitations}
+
+% TODO: memory usage discussion: the impact of lines and how arcs might help. or work graphs, or maybe mesh shaders. Limitation: memory must be allocated up-front. bump estimation, violates the no pre-curve-evaluation criterium but simpler techniques (like Wang's) exist and we use it but way over-estimates. This is a limitation of the GPU execution model. Explain this trade-off.
+  % -> TODO: Memory usage graphs for input scene buffers and LineSoup.
+
+\section{Performance}
+
+% TODO: We integrated our algorithm into a fully GPU-driven rendering pipeline, Vello WGSL
+% TODO: 3 alternative implementations: cubic->ES (for load balancing eval), cubic->ES->LineSoup (current), cubic->ES->ArcSoup
```

Comment:
Big TODO, but I'd love to see it :)

I think we should also have a contingency plan for landing the paper even without these measurements.

Reply:
I agree, I think we can drop the second TODO without losing much on the narrative. As discussed yesterday, I'll prioritize the cubic->ArcSoup experiment but it's not the end of the world if we don't get the GPU numbers for it.

Also a reminder for me to ablate the bbox atomic and measure without it.